### DISCUSSION TITLE
2021-11-29 - Frontier Tools Update

### DISCUSSION THREAD

**Discussion by MaximusHaximus** [near/community#137] [Archived]
The Frontier Tools team is working hard on near-api-js, the developer dashboard, and a Javascript SDK for smart contract development!

## [Javascript Smart Contract SDK](https://near.zulipchat.com/#narrow/stream/295801-nearinc.2Fdeveloper-platform/topic/SDK.20JS)
- @ailisp has been making a ton of progress on this! He has created a proof of concept that does the most expensive operation from our NFT smart contract and verified that it performs sufficiently that the contract doesn't run out of gas doing this operation!  This is a key discovery that validates our beliefs that this can work, especially if the JS runtime were to be bundled into nearcore on the nodes, rather than being compiled as user-space code into every JS based smart contract :)
- We're now at the stage where we are now talking about the JS side API designs!
- Our initial intention is to follow the Rust SDK's basic API everywhere that it makes sense, leveraging the added functionality that TS provides over AssemblyScript to provide a similar API to the one that the Rust SDK provides.
- We intend to only modify the APIs where it makes sense for idiomatic Javascript or where we have an opportunity to improve upon the user experience and eliminate confusing boilerplate.
- We're hoping to be able to allow cross-contract calls to be `await`'d idiomatically without forcing the user to create `.then()` style chains for the logic.  The tricky part of allowing this type of structure is dealing with the `await` keyword, but we think we may be able to allow idiomatic async JS syntax when making cross contract calls and doing other operations that create 'near promises'.  In particular, iteration that dispatches NEARPromises will need to be encapsulated for us to facilitate this.  We're hoping we won't need developers to nest code that needs to run based on the result of cross-contract calls (or other NEARPromise returning functions) in .then() chains!
- Discussion around bundling has begun! We don't want QuickJS to be loading modules!  We're considering making our JS SDK toolchain use [rollup.js](https://rollupjs.org/) which will give us a single source file from a group of modules _and_ provide tree-shaking that will remove dead code from the bundle that is used to build the smart contract code!

## [near-api-js](https://github.com/near/near-api-js/)
- [Adding support for RPC API keys](https://github.com/near/near-api-js/pull/729) (pending)
- [Adding ESM builds for native import usage (blocked)](https://github.com/near/near-api-js/pull/712) - it turns out this is a double-edged sword because once you use ESM imports, _all of the dependencies of your package must also support ESM imports :/_.  We haven't figured out the right solution here yet but it's in-flight!
- Adding some safeguards to avoid accidental corruption by someone deploying a multisig/2FA contract to an account that _already has a contract deployed to it_. Thanks for the cross-project support @Patrick1904  :)

## Developer Dashboard
- Work continues on the frontend for initial release!
- Discussion around contract statistics is ongoing
- Michael demoed the initial frontend in our all-hands! Nice work, @mpeterdev :)
- Basic monitoring is being implemented (StatsD or similar) for the initial release
- Also looking at adding basic vulnerability and license auditing for dependencies of the dev dashboard codebase for the initial release.

**Comment by maxzaver** [comment]
> This is a key discovery that validates our beliefs that this can work, especially if the JS runtime were to be bundled into nearcore on the nodes, rather than being compiled as user-space code into every JS based smart contract

@ailisp  @bowenwang1996 @MaximusHaximus  @matklad  FYI. It is very unlikely that we will compile JS runtime into nearcore any time in the future. Unless you find a way to isolate it into a hermetic (no side-effects) deterministic host function which is also easy to meter.

> The tricky part of allowing this type of structure is dealing with the await keyword

This is relevant: https://near.zulipchat.com/#narrow/stream/295558-nearinc.2Fprotocol-core/topic/Crazy.20idea.20to.20make.20cross-contact.20calls.20ergonomic/near/263370823

**Reply by ailisp** [reply]
Will try hard to shrink quickjs size then!

**Reply by bowenwang1996** [reply]
What is the current performance of running a JS runtime as a smart contract?

**Reply by ailisp** [reply]
The gas spent is good. with recent gas cost update, nft contract vm alone is about 1.x T gas. In a local node:
```
transaction_outcome.outcome.gas_burnt = 2427983606152
receipts_outcome[0].outcome.gas_burnt = 4373805221980
receipts_outcome[1].outcome.gas_burnt = 223182562500
```
Totally 7T gas.

Contract size: **527K** (almost the quickjs base size. code in js add very little to the number)

**Reply by bowenwang1996** [reply]
@ailisp is my understanding correct that with the current approach, we always "compile" a copy of js runtime into the wasm binary that is deployed on chain?

**Reply by ailisp** [reply]
Yes

**Reply by bowenwang1996** [reply]
@ailisp that seems a bit wasteful. Would be ideal if we only need to deploy the javascript runtime once on chain :) I think that may be possible if we have both https://github.com/near/NEPs/issues/93 and https://github.com/near/NEPs/discussions/173 implemented
